### PR TITLE
Simon/code refactor

### LINKED ISSUE #6: Make the code uniform across the different schemes
The current implementation makes use of different types and traits, and is quite messy.
The purpose is to reorganize the implementation getting rid of redundancy, improving readability and facilitating future implementations.

### PR DESCRIPTION
**Pull request by SimonRastikian**
This part concerns code refactoring from PR #15.
Closes #6

### CODE REVIEW

File: src/ecdsa/ot_based_ecdsa/triples/mta.rs [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+    let wait1 = chan.next_waitpoint();
+    let chi1 = SerializableScalar::<C>(chi1);
+    chan.send(wait1, &(chi1, seed));
+
+    Ok(beta)
+}
+
+/// Run the multiplicative to additive protocol
+#[allow(dead_code, clippy::type_complexity)]
+fn run_mta(
```
Comment (kevindeforth) [inline src/ecdsa/ot_based_ecdsa/triples/mta.rs:129]: This code is private and dead, lets just move inside the `test` module.

File: src/ecdsa/ot_based_ecdsa/triples/mta.rs [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    crypto::proofs::strobe_transcript::TranscriptRng,
+    protocol::{
+        internal::{make_protocol, PrivateChannel},
+        run_two_party_protocol, Participant, ProtocolError,
+    },
+};
+
+use crate::ecdsa::{Scalar, Secp256K1Sha256};
+
+type C = Secp256K1Sha256;
```
Comment (kevindeforth) [inline src/ecdsa/ot_based_ecdsa/triples/mta.rs:18]: Readability of this file might improve if we called this `Secp256` instead of `C`. When I read `C`, I expect a trait, not a struct.

File: src/ecdsa/ot_based_ecdsa/triples/mta.rs [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+/// The sender for multiplicative to additive conversion.
+pub async fn mta_sender(
+    mut chan: PrivateChannel,
+    v: Vec<(Scalar, Scalar)>,
+    a: Scalar,
+) -> Result<Scalar, ProtocolError> {
+    let size = v.len();
+
+    // Step 1
```
Comment (kevindeforth) [inline src/ecdsa/ot_based_ecdsa/triples/mta.rs:41]: Without context, `Step 1`, `Step 2`, ... are not very useful comments. It would be much more helpful if the comments explained what was done.

If there is some pseudo-code of this protocol, explaining what the individual steps do, then it would be good to reference it here.

**Review by gilcu3 — CHANGES_REQUESTED** [review]

File: src/ecdsa/ot_based_ecdsa/triples/mod.rs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+mod correlated_ot_extension;
+mod generation;
+mod mta;
+mod multiplication;
+mod random_ot_extension;
+
+pub use generation::{generate_triple, generate_triple_many, TripleGenerationOutput};
+
+#[cfg(test)]
+pub(crate) mod test;
```
Comment (kevindeforth) [inline src/ecdsa/ot_based_ecdsa/triples/mod.rs:75]: module references should be at the top of the file.

**Review by kevindeforth — COMMENTED** [review]
Partial review. Some things that I noticed:
- some of the protocols have no reference pseudocode (e.g. `ecdsa/ot_based_ecdsa/triples/mta`)
- quite a bit of dead code.

File: src/crypto/polynomials.rs [hunk 5]
```diff
@@ -0,0 +1,527 @@
...
+
+/// Computes the lagrange coefficient using a set of given points
+/// lamda_i(x) = \prod_j (x - x_j)/(x_i - x_j)
+/// where j != i
+/// If x is None then consider it as 0
+pub fn compute_lagrange_coefficient<C: Ciphersuite>(
+    points_set: &[Scalar<C>],
+    i: &Scalar<C>,
+    x: Option<&Scalar<C>>,
+) -> Result<SerializableScalar<C>, ProtocolError> {
```
Comment (kevindeforth) [inline src/crypto/polynomials.rs:344]: There is a mismatch between the comment and the function signature: `i` is used as an index in the comment, but is a scalar in the signature.

Maybe this might be an improvement:
```
Returns the value of the Larange basis polynomial in `x`: `\prod_j (x - x_j)/(x_i - x_j)`, where `j != i` and where `x_j` are elements in `point_set`.
Note that this function returns an error in case `x_j` is not in `point_set`.
Note that `point_set` must not have repeating values.
```

File: src/crypto/polynomials.rs [hunk 6]
```diff
@@ -0,0 +1,527 @@
...
+            .expect("coefficients must have at least one element")
+    }
+}
+
+/// Computes the lagrange coefficient using a set of given points
+/// lamda_i(x) = \prod_j (x - x_j)/(x_i - x_j)
+/// where j != i
+/// If x is None then consider it as 0
+pub fn compute_lagrange_coefficient<C: Ciphersuite>(
+    points_set: &[Scalar<C>],
```
Comment (kevindeforth) [inline src/crypto/polynomials.rs:341]: Does the `Scalar` interface allow passing an actual set here, to avoid duplicates? If so, lets do it. Otherwise, we might need to add a check inside the function.
If, for performance reasons, we want to avoid the check and not construct a set, then lets append an `_unchecked` suffix to the function name.

File: src/crypto/polynomials.rs [hunk 7]
```diff
@@ -0,0 +1,527 @@
...
+        PolynomialCommitment::new(coefficients)
+            .expect("coefficients must have at least one element")
+    }
+}
+
+/// Computes the lagrange coefficient using a set of given points
+/// lamda_i(x) = \prod_j (x - x_j)/(x_i - x_j)
+/// where j != i
+/// If x is None then consider it as 0
+pub fn compute_lagrange_coefficient<C: Ciphersuite>(
```
Comment (kevindeforth) [inline src/crypto/polynomials.rs:340]: lets add a test for this function. Specifically, for the edge cases.
Do we want to panic in case there are duplicate values?

File: src/crypto/polynomials.rs [hunk 8]
```diff
@@ -0,0 +1,527 @@
...
+            Ok(PolynomialCommitment { coefficients })
+        }
+    }
+}
+
+impl<C: Ciphersuite> Add for &PolynomialCommitment<C> {
+    type Output = PolynomialCommitment<C>;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        // zip iterates over the smaller vec
```
Comment (kevindeforth) [inline src/crypto/polynomials.rs:312]: Lets extract the degree of the output polynomial and reserve at the beginning of the function:
```Rust
let max_len = self.coefficients.len().max(rhs.coefficients.len());
let mut coefficients = Vec::with_capacity(max_len);ax_len);
```

File: src/crypto/polynomials.rs [hunk 9]
```diff
@@ -0,0 +1,527 @@
...
+        coeffcommitment.extend(self.get_coefficients());
+        PolynomialCommitment::new(coeffcommitment)
+            .expect("coefficients must have at least one element")
+    }
+
+    /// Set the constant value of this polynomial to a new group element
+    /// Aborts if the output polynomial is the identity
+    pub fn set_constant(&mut self, v: CoefficientCommitment<C>) -> Result<(), ProtocolError> {
+        if self.coefficients.len() == 1 && v.value() == C::Group::identity() {
+            return Err(ProtocolError::EmptyOrZeroCoefficients);
```
Comment (kevindeforth) [inline src/crypto/polynomials.rs:265]: same question here, why don't we allow the user to set the coefficent to zero?

File: src/crypto/commit.rs [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+        let mut hasher = Sha256::new();
+        hasher.update(COMMIT_LABEL);
+        hasher.update(r.as_ref());
+        hasher.update(b"start data");
+        rmp_serde::encode::write(&mut hasher, val)?;
+        Ok(Commitment(hasher.finalize().into()))
+    }
+
+    /// Check that a value and a randomizer match this commitment.
+    pub fn check<T: Serialize>(&self, val: &T, r: &Randomizer) -> Result<bool, Error> {
```
Comment (kevindeforth) [inline src/crypto/commit.rs:30]: lets add a test for this and the `commit` function below.

File: src/crypto/proofs/dlog.rs [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+
+use super::strobe_transcript::Transcript;
+use rand_core::CryptoRngCore;
+
+/// The label we use for hashing the statement.
+const STATEMENT_LABEL: &[u8] = b"dlog proof statement";
+/// The label we use for hashing the first prover message.
+const COMMITMENT_LABEL: &[u8] = b"dlog proof commitment";
+/// The label we use for generating the challenge.
+const CHALLENGE_LABEL: &[u8] = b"dlog proof challenge";
```
[the remark this answers is not in the document]
Comment (SimonRastikian) [reply src/crypto/proofs/dlog.rs:12]: Not really. Why should all constant strings start with Near?

File: src/ecdsa/ot_based_ecdsa/triples/mta.rs [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+    run_two_party_protocol(
+        s,
+        r,
+        &mut make_protocol(ctx_s.clone(), mta_sender(ctx_s.private_channel(s, r), v, a)),
+        &mut make_protocol(
+            ctx_r.clone(),
+            mta_receiver(ctx_r.private_channel(r, s), tv, b),
+        ),
+    )
+}
```
[the remark this answers is not in the document]
Comment (SimonRastikian) [reply src/ecdsa/ot_based_ecdsa/triples/mta.rs:147]: It's not dead code but I see the function should be defined in the unit test. Moved!

File: src/generic_dkg.rs [hunk 13]
```diff
@@ -494,27 +460,44 @@ async fn do_keyshare<C: Ciphersuite>(
             &all_hash_commitments,
```
[the remark this answers is not in the document]
Comment (SimonRastikian) [reply src/generic_dkg.rs:460]: If I recall well, most of the changes are indeed refactor.
Although I did extra things namely:
1) making one of the check we make late to slightly earlier (can only improve the security because if the check fails then the protocol will abort early)
2) Simplify the setup by using polynomials instead of arrays of scalars

File: src/crypto/README.md
[the remark this answers is not in the document]
Comment (kevindeforth) [reply src/crypto/README.md:1]: @SimonRastikian you labeled this as resolved, but the file is still here and in the same form it was before.

[the remark this answers is not in the document]
Comment (kevindeforth) [reply src/crypto/README.md:1]: (sorry, my browser / github was acting weird, this seems resolved?)

File: src/crypto/proofs/dlog.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (gilcu3) [reply src/crypto/proofs/dlog.rs:12]: That's exactly the question. I think it would be better to handle constants in a more centralized way, so that the function of each is easier to understand, and harder to confuse. Currently there are constants in many places, and for example why some constants contain "Near" and others do not is not so clear for the person reading the code.

File: src/generic_dkg.rs [hunk 14]
```diff
@@ -403,15 +359,15 @@ async fn do_keyshare<C: Ciphersuite>(
...
+    let mut all_full_commitments = ParticipantMap::new(&participants);
     let mut domain_separator = 0;
     // Make sure you do not call do_keyshare with zero as secret on an old participant
     let (old_verification_key, old_participants) =
         assert_keyshare_inputs(me, &secret, old_reshare_package)?;
 
     // Start Round 0
     let mut my_session_id = [0u8; 32]; // 256 bits
-    OsRng.fill_bytes(&mut my_session_id);
+    rng.fill_bytes(&mut my_session_id);
```
[the remark this answers is not in the document]
Comment (SimonRastikian) [reply src/generic_dkg.rs:370]: Ok then you are suggesting that I do not even put any OsRng as input to any function in generic_dkg?
What would you think would be a benefit in passing such arguments?

File: src/protocol/mod.rs [hunk 15]
```diff
@@ -49,16 +58,32 @@ impl fmt::Display for ProtocolError {
...
             ProtocolError::AssertionFailed(e) => write!(f, "assertion failed {e}"),
             ProtocolError::DKGNotSupported => write!(f, "the ciphersuite does not support DKG"),
+            ProtocolError::EmptyOrZeroCoefficients => {
+                write!(f, "Found empty polynomials or zero polynomial.")
+            }
             ProtocolError::ErrorExtractVerificationKey => write!(
                 f,
                 "could not extract the verification Key from the commitment."
             ),
+            ProtocolError::ErrorReducingBytesToScalar => write!(
```
[the remark this answers is not in the document]
Comment (SimonRastikian) [reply src/protocol/mod.rs:68]: Agree!
Let's keep things as they are in this PR and merge the next one.

File: src/ecdsa/ot_based_ecdsa/triples/batch_random_ot.rs [hunk 16]
```diff
@@ -294,7 +322,7 @@ pub async fn batch_random_ot_receiver_many<C: CSCurve, const N: usize>(
 
 /// Run the batch random OT protocol between two parties.
 #[allow(dead_code)]
```
[the remark this answers is not in the document]
Comment (SimonRastikian) [reply src/ecdsa/ot_based_ecdsa/triples/batch_random_ot.rs:324]: Used only in unit test, moved there.

File: src/ecdsa/ot_based_ecdsa/triples/mta.rs (hunk 3 above)
Comment (SimonRastikian) [reply src/ecdsa/ot_based_ecdsa/triples/mta.rs:37]: I agree. These steps are the same as the same as in the docs folder but I would want to change these comments once we tackle #5

File: src/ecdsa/ot_based_ecdsa/triples/test.rs [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+    protocol::Participant,
+};
+
+/// Create a new triple from scratch.
+///
+/// This can be used to generate a triple if you then trust the person running
+/// this code to forget about the values they generated.
+/// We prevent users from using it in non-testing env and attribute it to #[cfg(test)]
+#[cfg(test)]
+pub fn deal(
```
[the remark this answers is not in the document]
Comment (kevindeforth) [reply src/ecdsa/ot_based_ecdsa/triples/test.rs:20]: I know what a trusted dealer is. "dealing" refers to what happens to the output of this function.

A good function name usually reflects what the function does, not what is supposed to happen with its output.

File: src/participants.rs [hunk 18]
```diff
@@ -83,44 +83,18 @@ impl ParticipantList {
...
-            let q_scalar = q.scalar::<C>();
-            top *= q_scalar;
-            bot *= q_scalar - p_scalar;
-        }
-
-        top * bot.invert().unwrap()
-    }
-
-    /// Get the lagrange coefficient for a participant, relative to this list.
+    /// The lagrange coefficient are evaluated to zero
```
[the remark this answers is not in the document]
Comment (kevindeforth) [reply src/participants.rs:86]: This is probably due to my lack of background knowledge, but I genuinely don't understand what the comment is saying.

Specifically, what does it mean to "evaluate a coefficent to zero"?

File: src/ecdsa/ot_based_ecdsa/triples/test.rs [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+use crate::protocol::{
+    internal::{make_protocol, Comms},
+    run_two_party_protocol, Participant, ProtocolError,
+};
+
+/// Create a new triple from scratch.
+///
+/// This can be used to generate a triple if you then trust the person running
+/// this code to forget about the values they generated.
+/// We prevent users from using it in non-testing env and attribute it to #[cfg(test)]
```
Comment (kevindeforth) [inline src/ecdsa/ot_based_ecdsa/triples/test.rs:19]: stale comment

File: src/ecdsa/ot_based_ecdsa/triples/mta.rs (hunk 3 above)
Comment (kevindeforth) [reply src/ecdsa/ot_based_ecdsa/triples/mta.rs:37]: (I added a reference to this comment)

File: src/ecdsa/ot_based_ecdsa/triples/test.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (SimonRastikian) [reply src/ecdsa/ot_based_ecdsa/triples/test.rs:20]: Would you validate the name `deal_multiplicative_triples`?

File: src/crypto/polynomials.rs (hunk 6 above)
Comment (SimonRastikian) [reply src/crypto/polynomials.rs:341]: No, Scalar does not provide such a functionality.
One could try to do it internally but it I guess it is quite costly considering the number of times we actually use this function in a protocol.
It is possible to add  _unchecked in the name but not sure how much this suffix is expressive on which check we are talking about...

Comment (SimonRastikian) [reply src/crypto/polynomials.rs:341]: I would go with either Option (3) or nothing,

File: src/crypto/polynomials.rs (hunk 7 above)
Comment (SimonRastikian) [reply src/crypto/polynomials.rs:340]: The function is tested indirectly in eval_exponent_interpolation and in eval_interpolation.
I do not see any other way in testing it? If you suggest something I am open to add it.
Isn't the duplicates question the same as the previous comment?

File: src/ecdsa/ot_based_ecdsa/triples/generation.rs [hunk 20]
```diff
@@ -1,42 +1,55 @@
...
-use crate::protocol::internal::Comms;
+use super::{
+    multiplication::{multiplication, multiplication_many},
+    TriplePub, TripleShare,
+};
+
+/// Encode an arbitrary serializable value into a vec.
+fn encode<T: Serialize>(val: &T) -> Vec<u8> {
+    rmp_serde::encode::to_vec(val).expect("failed to encode value")
+}
```
[the remark this answers is not in the document]
Comment (SimonRastikian) [reply src/ecdsa/ot_based_ecdsa/triples/generation.rs]: Please check [e0d2540](https://github.com/near/threshold-signatures/pull/22/commits/e0d2540ec47bc0222624ffbb996ff9105956eab9) and tell me what you think.

I just found that most of the serialization can panic (mostly due to legacy functions). It would be great @gilcu3  if you could open an issue about it and handle this problem yourself as I would like to focus more on pushing forward with the core features of robust ecdsa and tests generalizations.

[the remark this answers is not in the document]
Comment (gilcu3) [reply src/ecdsa/ot_based_ecdsa/triples/generation.rs]: I think the refactoring in [e0d2540](https://github.com/near/threshold-signatures/pull/22/commits/e0d2540ec47bc0222624ffbb996ff9105956eab9) indeed makes the original comment not relevant anymore.

> I just found that most of the serialization can panic (mostly due to legacy functions).

If you open a (small/quick) issue with an example of how it fails currently I can expand it if necessary and then take care of it later. That should be the most efficient, as currently I don't see how is the panic happening.

File: src/participants.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (near-bookrock) [reply src/participants.rs]: Btw, the docs are just plain markdown. Would be good to link to a wiki in cases like these. I think most people on the team would not catch this.

[the remark this answers is not in the document]
Comment (near-bookrock) [reply src/participants.rs]: I understand. Just in case it wasn't clear what I mean, I mean that you can copy paste links like these into the code and the rust doc will be rendered as markdown.

```rust
    /// The [lagrange coefficient](https://en.wikipedia.org/wiki/Lagrange_polynomial) are evaluated to zero
    /// Use generic frost library types
    pub fn generic_lagrange<C: Ciphersuite>(&self, p: Participant) -> Scalar<C> {
```

**Review by kevindeforth — COMMENTED** [review]
Sorry, not done yet. This is like eating an elephant.

I don't think I found big issues, but there are two things that give me pause in the code I saw thus far:
- lack of test coverage (left a comment with the polynomials.rs file)
- we have quite a few panics in low-level functions. Would it be possible to return errors instead?

File: src/crypto/proofs/dlog.rs [hunk 21]
```diff
@@ -0,0 +1,162 @@
...
+        .as_ref()
+        .len();
+    // Serializing the identity might fail!
+    // this is a workaround to be able to serialize even this infinity point.
+    let ser = match <<C as frost_core::Ciphersuite>::Group as Group>::Serialization::try_from(vec![
+            0u8;
+            size
+        ]) {
+        Ok(ser) => ser,
+        _ => panic!("Should not raise error"),
```
[the remark this answers is not in the document]
Comment (near-bookrock) [reply src/crypto/proofs/dlog.rs:79]: Good observations. **There should be no panics in library code.**

**Review by kevindeforth — COMMENTED** [review]
The more I read, the less need I see for panics in this code.

In some functions inside crypto/proofs, panics can be avoided altogheter by writing the code differently.
In other parts, it seems like the top-level function is returning a `Result`, so we could easily propagate the error to the caller instead of crashing the program.

Some panics would get triggered if the random number generator returned a specific element (e.g. identity). While highly unlikely to actually happen, this is not a good reason to panic, because it could be easily avoided (by sampling until we find a non-identity element).

File: src/crypto/polynomials.rs [hunk 22]
```diff
@@ -0,0 +1,548 @@
...
+    /// Used usually after sending a smaller polynomial to prevent serialization from
+    /// failing if the constant term is the identity
+    pub fn extend_with_zero(&self) -> Self {
+        let mut coeffcommitment = vec![<C::Group as Group>::Field::zero()];
+        coeffcommitment.extend(self.get_coefficients());
+        Polynomial::new(coeffcommitment).expect("coefficients must have at least one element")
+    }
+}
+
+/******************* Polynomial Commitment *******************/
```
[the remark this answers is not in the document]
Comment (SimonRastikian) [reply src/crypto/polynomials.rs:167]: I would'nt prefer to do that as I see both struct as two sides of the same coin. They both behave similarly, provide similar interfaces and make use of a common function `compute_lagrange_coefficient`

File: src/crypto/polynomials.rs [hunk 23]
```diff
@@ -0,0 +1,548 @@
...
+        self.eval_on_point(id)
+    }
+
+    /// Computes polynomial interpolation on a specific point
+    /// using a sequence of sorted elements
+    /// Input requirements:
+    ///     * identifiers MUST be pairwise distinct and of length greater than 1
+    ///     * shares and identifiers must be of same length
+    // Returns error if shares' and identifiers' lengths are distinct or less than or equals to 1
+    pub fn eval_interpolation(
```
[the remark this answers is not in the document]
Comment (SimonRastikian) [reply src/crypto/polynomials.rs:111]: The correct behaviour is already tested in `poly_generate_evaluate_interpolate` and `com_generate_evaluate_interpolate`. I will write a test that fails on wrong inputs.

File: src/crypto/proofs/dlog.rs (hunk 21 above)
[the remark this answers is not in the document]
Comment (SimonRastikian) [reply src/crypto/proofs/dlog.rs:79]: Comment added in #31

File: src/crypto/proofs/dlogeq.rs [hunk 24]
```diff
@@ -0,0 +1,212 @@
...
+#[serde(bound = "C: Ciphersuite")]
+pub struct Proof<C: Ciphersuite> {
+    e: SerializableScalar<C>,
+    s: SerializableScalar<C>,
+}
+
+/// Encodes two EC points into a vec including the identity point.
+/// Should be used with HIGH precaution as it allows serializing the identity point
+/// deviating from the standard
+fn encode_two_points<C: Ciphersuite>(point_1: &Element<C>, point_2: &Element<C>) -> Vec<u8> {
```
[the remark this answers is not in the document]
Comment (SimonRastikian) [reply src/crypto/proofs/dlogeq.rs]: Cleaned in [0d93379](https://github.com/near/threshold-signatures/pull/22/commits/0d93379c1efda415aa7681ca1811e77f4addd121)

File: src/crypto/polynomials.rs [hunk 25]
```diff
@@ -0,0 +1,548 @@
...
+            return Err(ProtocolError::InvalidInterpolationArguments);
+        }
+
+        // Compute the Lagrange coefficients
+        for (id, share) in identifiers.iter().zip(shares) {
+            let lagrange_coefficient = compute_lagrange_coefficient::<C>(identifiers, id, point)?;
+
+            // Compute y = f(point) via polynomial interpolation of these points of f
+            interpolation = interpolation + (lagrange_coefficient.0 * share.0);
+        }
```
[the remark this answers is not in the document]
Comment (SimonRastikian) [reply src/crypto/polynomials.rs:128]: No, not as far as I know.
We can only get better understanding of the full performance once we handle Task #8

File: src/participants.rs
[the remark this answers is not in the document]
Comment (SimonRastikian) [reply src/participants.rs:1]: Could you develop on this so I can take proper notes?

File: src/generic_dkg.rs [hunk 26]
```diff
@@ -434,31 +389,37 @@ async fn do_keyshare<C: Ciphersuite>(
...
     let commitment_hash = domain_separate_hash(domain_separator, &(&me, &commitment, &session_id));
     let wait_round_1 = chan.next_waitpoint();
     chan.send_many(wait_round_1, &commitment_hash);
     // receive commitment_hash
+    let mut seen = ParticipantCounter::new(&participants);
     let mut all_hash_commitments = ParticipantMap::new(&participants);
     all_hash_commitments.put(me, commitment_hash);
-    while !all_hash_commitments.full() {
+    seen.put(me);
+    while !seen.full() {
```
[the remark this answers is not in the document]
Comment (SimonRastikian) [reply src/generic_dkg.rs]: [fac3da4](https://github.com/near/threshold-signatures/pull/22/commits/fac3da4a7e86a48ac8bd9777493c56e57d4d42ed)

File: src/generic_dkg.rs [hunk 27]
```diff
@@ -421,10 +373,13 @@ async fn do_keyshare<C: Ciphersuite>(
...
     domain_separator += 1;
-    let secret_coefficients = generate_secret_polynomial::<C>(secret, threshold, &mut rng);
+    // the degree of the polynomial is threshold - 1
+    let secret_coefficients =
+        Polynomial::<C>::generate_polynomial(Some(secret), threshold - 1, &mut OsRng)?;
 
     // Compute the multiplication of every coefficient of p with the generator G
-    let coefficient_commitment = generate_coefficient_commitment::<C>(&secret_coefficients);
+    let coefficient_commitment = generate_coefficient_commitment::<C>(&secret_coefficients)
+        .expect("expects non constant polynomial (i.e. threshold strictly larger than 1)");
```
[the remark this answers is not in the document]
Comment (SimonRastikian) [reply src/generic_dkg.rs]: cannot panic because the secret_coefficients are generated properly by the previous call. 
Anyway, I just propagated the error in [3eb2333](https://github.com/near/threshold-signatures/pull/22/commits/3eb2333ef1ea9885d40dc7ed76f28288a9afa5ff)

File: src/crypto/polynomials.rs [hunk 28]
```diff
@@ -0,0 +1,619 @@
...
+        rng: &mut impl CryptoRngCore,
+    ) -> Result<Self, ProtocolError> {
+        let poly_size = degree + 1;
+        let mut coefficients = Vec::with_capacity(poly_size);
+        // insert the secret share if exists
+        let secret = secret.unwrap_or_else(|| <C::Group as Group>::Field::random(rng));
+
+        coefficients.push(secret);
+        for _ in 1..poly_size {
+            coefficients.push(<C::Group as Group>::Field::random(rng));
```
[the remark this answers is not in the document]
Comment (SimonRastikian) [reply src/crypto/polynomials.rs:64]: Once again, this is cryptographically negligible to draw a zero, the probability is around 1/2^256

[the remark this answers is not in the document]
Comment (SimonRastikian) [reply src/crypto/polynomials.rs:64]: This probability is roughly the same as breaking collision resistance of sha512

File: src/ecdsa/ot_based_ecdsa/triples/batch_random_ot.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (netrome) [reply src/ecdsa/ot_based_ecdsa/triples/batch_random_ot.rs]: How is this used in the MPC repo if the function is private to the crate?

File: src/crypto/polynomials.rs
[the remark this answers is not in the document]
Comment (netrome) [reply src/crypto/polynomials.rs:1]: My take here:

1. Adding simple tests for the methods you outlined is a reasonable thing to do before merging.
2. Arguing about not implementing the trivial tests will take longer time than just implementing them[^1].

[^1]: This non-arguing reasoning applies only to "lazyness argument". I.e. "I agree this is good, but let's not do it now". Any other argument such as "For reason X, Y and Z these tests actually make the code base worse" would still be valid, but I don't expect such arguments for these test.

File: src/crypto/proofs/dlog.rs
[the remark this answers is not in the document]
Comment (netrome) [reply src/crypto/proofs/dlog.rs:1]: > I do not see how to test any of the rest individually. Could you please help me find something?

Shallow thought: One variant would be to test against snapshot values. I.e. record a valid proof, and make a test that the prove function outputs the exact same proof for the same parameters. Secondly make a unit test that verification passes for this proof, and an additional one where verification fails for a randomly generated proof.

File: src/crypto/polynomials.rs (hunk 7 above)
Comment (netrome) [reply src/crypto/polynomials.rs:352]: > I do not see any other way in testing it?

What is the problem with adding a simple unit test where you feed some pre-determined values and assert some expected output?

I'd expect test cases covering:
1. Happy case with some simple values.
2. x: None.
3. x_i not in point set.
4. Duplicate values in point set.

Is there any challenges associated with implementing these test cases?

File: src/crypto/polynomials.rs
[the remark this answers is not in the document]
Comment (SimonRastikian) [reply src/crypto/polynomials.rs:1]: Completed in [82d479b](https://github.com/near/threshold-signatures/pull/22/commits/82d479b1d7245646dc4b91543164cb10beabddd0)

File: src/generic_dkg.rs [hunk 29]
```diff
@@ -421,10 +373,13 @@ async fn do_keyshare<C: Ciphersuite>(
...
     // generate a proof of knowledge if the participant me is not holding a secret that is zero
     let proof_domain_separator = domain_separator;
     let proof_of_knowledge = compute_proof_of_knowledge(
```
[the remark this answers is not in the document]
Comment (kevindeforth) [reply src/generic_dkg.rs:385]: Oh, I mis-read the if condition in the function.

Then it's even simpler:
```rust
    // I am allowed to send none only if I am a new participant
    let generate_proof: bool = if let Some(old) = &old_participants {
        old.contains(me)
    } else {
        true
    };
    let proof_of_knowledge = if generate_proof {
        Some(proof_of_knowledge(
            &session_id,
            domain_separator,
            me,
            &secret_coefficients,
            &coefficient_commitment,
        )?)
    } else {
        None
    };
```

[the remark this answers is not in the document]
Comment (SimonRastikian) [reply src/generic_dkg.rs]: Done in  [c31c8d3](https://github.com/near/threshold-signatures/pull/22/commits/c31c8d31f909b675988b44c586d84016ce006aca)

File: src/ecdsa/ot_based_ecdsa/triples/test.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (netrome) [reply src/ecdsa/ot_based_ecdsa/triples/test.rs:20]: Yeah this is not a blocker for the PR. Let's not bikeshed this too much, thanks for filling me in. If you have more time I would love a primer on the triples or just a link to some good resources. Otherwise I'm fine as well. My limited knowledge is that they help perform multi-party multiplications but I don't recall much else.

File: src/ecdsa/ot_based_ecdsa/triples/batch_random_ot.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (SimonRastikian) [reply src/ecdsa/ot_based_ecdsa/triples/batch_random_ot.rs]: @netrome not sure what you're asking here. But if it is about `run_batch_random_ot` function then this has been moved to test module. This code version is outdated  see https://github.com/near/threshold-signatures/pull/22/files#diff-828c7600c9ca003925b4117ddedc488adc75289ad104297547edefabda01be56R352-R378

File: src/crypto/polynomials.rs [hunk 30]
```diff
@@ -0,0 +1,829 @@
...
+        Ok(())
+    }
+
+    /// Extends the Polynomial with an extra value as a constant
+    /// Used usually after sending a smaller polynomial to prevent serialization from
+    /// failing if the constant term is the identity
+    pub fn extend_with_zero(&self) -> Self {
+        let mut coeffcommitment = vec![<C::Group as Group>::Field::zero()];
+        coeffcommitment.extend(self.get_coefficients());
+        Polynomial::new(coeffcommitment).expect("coefficients must have at least one element")
```
[the remark this answers is not in the document]
Comment (SimonRastikian) [reply src/crypto/polynomials.rs]: Done in [afb450b](https://github.com/near/threshold-signatures/pull/22/commits/afb450bf23ddeba6591349a39f67fab704ca62ce)

File: src/crypto/polynomials.rs [hunk 31]
```diff
@@ -0,0 +1,829 @@
...
+    }
+
+    /// Extends the Commited Polynomial with an extra value as a constant
+    /// Used usually after sending a smaller polynomial to prevent serialization from
+    /// failing if the constant term is the identity
+    pub fn extend_with_identity(&self) -> Self {
+        let mut coeffcommitment = vec![CoefficientCommitment::<C>::new(C::Group::identity())];
+        coeffcommitment.extend(self.get_coefficients());
+        PolynomialCommitment::new(coeffcommitment)
+            .expect("coefficients must have at least one element")
```
[the remark this answers is not in the document]
Comment (SimonRastikian) [reply src/crypto/polynomials.rs]: Done in [afb450b](https://github.com/near/threshold-signatures/pull/22/commits/afb450bf23ddeba6591349a39f67fab704ca62ce)

File: src/crypto/polynomials.rs [hunk 32]
```diff
@@ -0,0 +1,829 @@
...
+            .expect("coefficients must have at least one element set to non-zero")
+    }
+
+    /// Set the constant value of this polynomial to a new scalar
+    /// Abort if the output polynomial is zero
+    pub fn set_nonzero_constant(&mut self, v: Scalar<C>) -> Result<(), ProtocolError> {
+        if self.coefficients.len() == 1 && v == <C::Group as Group>::Field::zero() {
+            return Err(ProtocolError::EmptyOrZeroCoefficients);
+        }
+        self.coefficients[0] = v;
```
[the remark this answers is not in the document]
Comment (SimonRastikian) [reply src/crypto/polynomials.rs:163]: I find this is making the code extremely heavy. Check [290af98](https://github.com/near/threshold-signatures/pull/22/commits/290af984820021ffe9758a34cc14c5def01aece0)

File: src/crypto/polynomials.rs [hunk 33]
```diff
@@ -0,0 +1,829 @@
...
+        // fails only if:
+        // * polynomial is of degree 0 and the constant term is 0
+        // * polynomial degree is the max of usize, and so degree + 1 is 0
+        // such cases never happen in a classic (non-malicious) implementations
+        Self::new(coefficients)
+    }
+
+    /// Returns the constant term
+    pub fn eval_at_zero(&self) -> SerializableScalar<C> {
+        SerializableScalar(self.coefficients[0])
```
[the remark this answers is not in the document]
Comment (SimonRastikian) [reply src/crypto/polynomials.rs]: I find this is making the code extremely heavy. Check [290af98](https://github.com/near/threshold-signatures/pull/22/commits/290af984820021ffe9758a34cc14c5def01aece0)

File: src/crypto/polynomials.rs [hunk 34]
```diff
@@ -0,0 +1,829 @@
...
+            interpolation = interpolation + (share.value() * lagrange_coefficient.0);
+        }
+
+        Ok(CoefficientCommitment::new(interpolation))
+    }
+
+    /// Extends the Commited Polynomial with an extra value as a constant
+    /// Used usually after sending a smaller polynomial to prevent serialization from
+    /// failing if the constant term is the identity
+    pub fn extend_with_identity(&self) -> Self {
```
[the remark this answers is not in the document]
Comment (SimonRastikian) [reply src/crypto/polynomials.rs]: Done in [0e0e1d8](https://github.com/near/threshold-signatures/pull/22/commits/0e0e1d88985d176db75727d4ae555252712d97ac)

**Comment by netrome** [comment]
> Although I do not find it very clean to have a sending channel fail, I am unsure for the moment what is the best approach to deal with this.

Can you elaborate? Is this due to the fallible serialization being used in the send functions, and if so - would it be possible to formulate these in a non-fallible way?

File: src/ecdsa/ot_based_ecdsa/triples/test.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (kevindeforth) [reply src/ecdsa/ot_based_ecdsa/triples/test.rs:20]: > I'm curious about the security assumption for the trusted dealer. If this is a malicious party, how much damage can they make?

It is absolutely disastrous.
The values `a` and `b` are known to all participants. If the trusted dealer shares the values of the masks `λ` or `μ`, then the secrets `x` and `y` can be re-constructed.
